src/planner/binder/expression/bind_operator_expression.cpp

```diff
@@ -147,6 +149,16 @@ BindResult ExpressionBinder::BindExpression(OperatorExpression &op, idx_t depth)
 		}
 	}
 
+	if (op.type == ExpressionType::COMPARE_IN || op.type == ExpressionType::COMPARE_NOT_IN) {
+		auto result_type = children[0]->return_type;
+		for (idx_t i = 1; i < children.size(); ++i) {
+			result_type = BoundComparisonExpression::BindComparison(result_type, children[i]->return_type);
```

Comment:
This explicit type resolution here should not be necessary anymore, right? Can't we use `result_type` that we obtained from `ResolveOperatorType` directly here?

Reply:
> This explicit type resolution here should not be necessary anymore, right? Can't we use `result_type` that we obtained from `ResolveOperatorType` directly here?

I don't think so. The `ResolveOperatorType` will invoke `ResolveInType()` in this case, and finally invoke `LogicalType::MaxLogicalType()`. In this function, varchar collations is determined like this:
``` c++
	if (type_id == LogicalTypeId::VARCHAR) {
		// varchar: use type that has collation (if any)
		if (StringType::GetCollation(right).empty()) {
			return left;
		}
		return right;
	}
```

This part logic is very simple and there is no distinction between the types of collation. But for other operators with collations (for example, Comparison, Between), they'll do more check in function `BindComparison`. If the collations are not the same, an Exception will be raised.

Give a simple example:
``` sql
create table tbl (a varchar collate noaccent);
select * from tbl where a = 'abc' collate nocase; -- sql1
select * from tbl where a in ('abc' collate nocase); -- sql2
```
`sql1` will fail because left and right expression has different collations. I think sql2 should act the same as `sql1`, but without `BindComparison`, this will use nocase collation.

The above is based on the code reading because I didn't get my laptop right now. I can do the test this evening and post the result pictures.

Reply:
In that case I would indeed say the type resolution in `ResolveOperatorType` should change. Currently the type resolution is in essence happening twice which seems unnecessary.

Reply:
I think you are right. `ResolveOperatorType()` and `BindComparison()` both invoke `LogicalType::MaxLogicalType()`. Maybe replace the `LogicalType::MaxLogicalType()` with `BindComparison()` for IN/NOT IN operator. But `ResolveInType()` is also used for `COALESCE` operator, which is not comparison and shouldn't use `BindComparison`, I think.

Maybe add a new input parameter for `ResolveInType()` to determine whether use `LogicalType::MaxLogicalType()` or `BindComparison()`. What do you think?

Reply:
I agree with you.  There're several ideas after re-read the code:
1. Since `Cast` is done by `ResolveInType()`, I think `PushCollation()` also should be done by this function. `PushCollation()` need `ClientContext` as parameter, but `ResolveInType()` is a static function, so `context` have to be passed by input parameter.
2. `ResolveInType()` are only invoked for (1) `IN/NOT_IN` (2) `COALESCE`. `BoundComparisonExpression::BindComparison` will invoke `LogicalType::MaxLogicalType` at first, but some extra adjustments for `DECIMAL` and `VARCHAR` is done after that. Adjustments for `VARCHAR` is fine, but adjustments for `DECIMAL` are unnecessary for `COALESCE`, I think, which will modify the scale and width if other expressions are not `DECIMAL`. So `BindComparison` only works on `IN/NOT_IN` operator.
3. If `any_varchar` and `any_enum` are both true, `max_type` will be re-assigned and collation information will be lost. In case this happens, if `max_type` is already `VARCHAR`, skip the re-assignment.

Please let me know if the change is not appropriate.